File: app/src/main/java/org/sil/storyproducer/tools/AudioPlayer.java
package org.sil.storyproducer.tools;

import android.media.MediaPlayer;

import java.io.IOException;

public class AudioPlayer {

    MediaPlayer mPlayer;

    /**
     * Constructor for Audio Player, no params
     */
    public AudioPlayer() {
        mPlayer = new MediaPlayer();
    }

    /**
     * Only sets the path for the audio to
     * @param path String path for the audio
     */
    public void setPath(String path) {
        try {
            mPlayer.setDataSource(path);
        } catch (IOException e) {
            //TODO maybe something with this exception
            e.printStackTrace();
        }
        try {
            mPlayer.prepare();
        } catch (IOException e) {
            //TODO maybe something with this exception
            e.printStackTrace();
        }
    }

    /**
     * Plays the audio with the given path
     * @param path for the path where the audio resides
     */
    public void playWithPath(String path) {
        try {
            mPlayer.setDataSource(path);
        } catch (IOException e) {
            //TODO maybe something with this exception
            e.printStackTrace();
        }
        try {
            mPlayer.prepare();
        } catch (IOException e) {
            //TODO maybe something with this exception
            e.printStackTrace();
        }
        mPlayer.start();
    }

    /**
     * Pauses the audio if it is currenlty being played
     */
    public void pauseAudio() {
        if(mPlayer != null && mPlayer.isPlaying()) {
            try {
                mPlayer.pause();
            } catch (IllegalStateException e) {
                //TODO maybe something with this exception
                e.printStackTrace();
            }
        }
    }

    /**
     * Resumes the audio from where it was last paused
     */
    public void resumeAudio() {
        if(mPlayer != null) {
            int pauseSpot = mPlayer.getCurrentPosition();
            mPlayer.seekTo(pauseSpot);
            mPlayer.start();
        }
    }

    /**
     * Stops the audio if it is currenlty being played
     */
    public void stopAudio() {
        if(mPlayer!= null && mPlayer.isPlaying()) {
            try {
                mPlayer.stop();
            } catch (IllegalStateException e) {
                //TODO maybe something with this exception
                e.printStackTrace();
            }
        }
    }

    /**
     * Stops the audio and releases it if it is currenlty being played
     */
    public void releaseAudio() {
        if(mPlayer!= null && mPlayer.isPlaying()) {
            try {

                mPlayer.stop();
                mPlayer.release();
            } catch (IllegalStateException e) {
                //TODO maybe something with this exception
                e.printStackTrace();
            } finally {
                mPlayer = null;   //this set to null so that an error doesn't occur if someone trys to release audio again
            }
        }
    }

    /**
     * returns the duration of the audio as an int
     * @return the duration of the audio as an int
     */
    public int getAudioDurationInSeconds() {
        return (int)(mPlayer.getDuration() * 0.001);
    }

    /**
     * returns the duration of the audio as an int in miliseconds
     * @return the duration of the audio as an int
     */
    public int getAudioDurationInMilliseconds() {
        return (int)mPlayer.getDuration();
    }

    /**
     * Seeks to the parameter in milliseconds
     * @param msec milliseconds for where to seek to in the audio
     */
    public void seekTo(int msec) {
        mPlayer.seekTo(msec);
    }

    /**
     * sets the completion listener
     * @param listener
     */
    public void audioCompletionListener(MediaPlayer.OnCompletionListener listener) {
        mPlayer.setOnCompletionListener(listener);
    }

    /**
     * sets the volume of the audio
     * @param volume the float for the volume from 0.0 to 1.0
     */
    public void setVolume(float volume) {
        mPlayer.setVolume(volume, volume);
    }

    /**
     * returns if the audio is being played or not
     * @return true or false based on if the audio is being played
     */
    public boolean isAudioPlaying() {
        return mPlayer.isPlaying();
    }
}

File: app/src/main/java/org/tyndalebt/spadv/tools/media/AudioPlayer.kt
package org.tyndalebt.spadv.tools.media

import android.content.Context
import android.media.MediaPlayer
import android.net.Uri
import org.tyndalebt.spadv.model.PhaseType
import org.tyndalebt.spadv.model.Workspace
import org.tyndalebt.spadv.tools.file.getStoryUri
import org.tyndalebt.spadv.model.WORD_LINKS_DIR

class AudioPlayer {

    private var mPlayer: MediaPlayer
    private var fileExists: Boolean = false
    private var onCompletionListenerPersist: MediaPlayer.OnCompletionListener? = null

    var currentPosition: Int
        get() =
            try{ mPlayer.currentPosition
            } catch (e : Exception){ 0 }
        set(value) {
            try {
                mPlayer.seekTo(value)
            } catch (e : Exception) {}
        }

    /**
     * returns the duration of the audio as an int in milliseconds
     * @return the duration of the audio as an int
     */
    val audioDurationInMilliseconds: Int
        get() = mPlayer.duration

    /**
     * returns if the audio is being played or not
     * @return true or false based on if the audio is being played
     */
    val isAudioPlaying: Boolean
        get() {
            try {
                return mPlayer.isPlaying
            } catch (e: IllegalStateException) {
                return false
            }

        }

    var isAudioPrepared: Boolean = false
        private set

    /**
     * Constructor for Audio Player, no params
     */
    init {
        mPlayer = MediaPlayer()
        fileExists = false
    }

    fun setSource(context: Context, uri: Uri) : Boolean {
        try {
            mPlayer.release()
            mPlayer = MediaPlayer()
            mPlayer.setOnCompletionListener(onCompletionListenerPersist)
            mPlayer.setDataSource(context, uri)
            fileExists = true
            isAudioPrepared = true
            mPlayer.prepare()
            currentPosition = 0
        } catch (e: Exception) {
            //TODO maybe do something with this exception
            fileExists = false
            isAudioPrepared = false
        }
        return fileExists
    }
    /**
     * set the audio file from the worskspace data
     * @return true if the file exists, false if it does not.
     */

    fun setStorySource(context: Context, relPath: String,
                       storyName: String = Workspace.activeStory.title) : Boolean {
        val uri: Uri = if (Workspace.activePhase.phaseType == PhaseType.WORD_LINKS){
            getStoryUri(relPath, WORD_LINKS_DIR) ?: return false
        }
        else{
            getStoryUri(relPath,storyName) ?: return false
        }
        return setSource(context, uri)
    }

    fun playAudio() {
        currentPosition = 0
        resumeAudio()
    }

    /**
     * Pauses the audio if it is currently being played
     */
    fun pauseAudio() {
        try {
            if(mPlayer.isPlaying)
                mPlayer.pause()
        } catch (e: Exception) {}
    }

    /**
     * Resumes the audio from where it was last paused
     */
    fun resumeAudio() {
        try {
            if(fileExists) {
                mPlayer.start()
            }
        } catch (e: Exception) { }

    }

    /**
     * Stops the audio if it is currently being played
     */
    fun stopAudio() {
        try {
            if(mPlayer.isPlaying) mPlayer.pause()
            if(currentPosition != 0) currentPosition = 0
        } catch (e: Exception) {}
    }

    /**
     * Releases the MediaPlayer object after completion
     */
    fun release() {
        isAudioPrepared = false
        try {
            mPlayer.release()
        } catch (e : Exception) {}
    }

    /**
     * Seeks to the parameter in milliseconds
     * @param msec milliseconds for where to seek to in the audio
     */
    fun seekTo(msec: Int) {
        if(!fileExists) return
        mPlayer.seekTo(msec)
    }

    /**
     * sets the completion listener
     * @param listener handler for OnCompletionListener
     */
    fun onPlayBackStop(listener: MediaPlayer.OnCompletionListener) {
        onCompletionListenerPersist = listener
        mPlayer.setOnCompletionListener(listener)
    }

    /**
     * sets the volume of the audio
     * @param volume the float for the volume from 0.0 to 1.0
     */
    fun setVolume(volume: Float) {
        mPlayer.setVolume(volume, volume)
    }

    companion object {

        private val TAG = "AudioPlayer"
    }
}
